web-api/src/storage/postgres.rs

```diff
@@ -372,6 +461,118 @@ impl Database {
+                    builder
+                        .reset()
+                        .push_tuple(ids.by_ref().take(Self::BIND_LIMIT))
+                        .build()
+                        .try_map(|row| DocumentId::from_row(&row))
+                        .fetch_all(&mut *tx)
+                        .await?,
+                );
+            }
+
+            Ok(candidates)
+        } else {
+            sqlx::query_as("SELECT document_id FROM document WHERE is_candidate = TRUE;")
+                .fetch_all(tx)
+                .await
+                .map_err(Into::into)
+        }
+    }
+
+    #[cfg(feature = "ET-4089")]
+    async fn set_candidates(
+        &self,
+        ids: impl IntoIterator<IntoIter = impl Clone + ExactSizeIterator<Item = &DocumentId>>,
+    ) -> Result<(Vec<IngestedDocument>, Vec<DocumentId>, Warning<DocumentId>), Error> {
+        let mut tx = self.pool.begin().await?;
+
+        let all = ids.into_iter();
+        let unchanged = Database::get_candidates(&mut tx, Some(all.clone())).await?;
+
+        let mut builder = QueryBuilder::new(
+            "UPDATE document
+            SET is_candidate = TRUE
+            WHERE is_candidate = FALSE AND document_id IN ",
+        );
+        let mut ingested = Vec::with_capacity(all.len());
+        let mut ids = all.clone().peekable();
+        while ids.peek().is_some() {
+            ingested.extend(
+                builder
+                    .reset()
+                    .push_tuple(ids.by_ref().take(Self::BIND_LIMIT))
+                    .push(" RETURNING document_id, snippet, properties, tags, embedding;")
+                    .build()
+                    .try_map(|row| {
+                        QueriedCandidateDocument::from_row(&row).map(|document| IngestedDocument {
+                            id: document.document_id,
+                            snippet: document.snippet,
+                            properties: document.properties.0,
+                            tags: document.tags,
+                            embedding: document.embedding,
+                            is_candidate: true,
+                        })
+                    })
+                    .fetch_all(&mut tx)
+                    .await?,
+            );
+        }
+
+        let mut builder = QueryBuilder::new(
+            "UPDATE document
+            SET is_candidate = FALSE
+            WHERE is_candidate = TRUE AND document_id NOT IN ",
+        );
+        let mut deleted = Vec::new();
+        let mut ids = all.clone().peekable();
+        while ids.peek().is_some() {
+            deleted.extend(
+                builder
+                    .reset()
+                    .push_tuple(ids.by_ref().take(Self::BIND_LIMIT))
+                    .push(" RETURNING document_id;")
+                    .build()
+                    .try_map(|row| DocumentId::from_row(&row))
+                    .fetch_all(&mut tx)
+                    .await?,
+            );
+        }
```

Comment:
:wrench: 

the unsetting part will not work as you can't split a `NOT IN` into multiple queies as on each split documents from the other splits will be "not in this split".

we probably could do `UPDATE document SET is_candidate = FALSE` followed by  a split `UPDATE document SET is_candidate = TRUE WHERE document_id IN (?...)`.

Besides that if the number of documents is much smaller then the number of documents which are candidates we probably should create a INDEX over all candidates and then delete all from index and set new through I don't know for sure I would need to check out how postges handles this kind of operations  performance wise.

Reply:
good point, damn bind limits. i'll add the index and update the query.

Reply:
i also changed the es delete to a retain, this simplifies the code overall and also keeps es in a cleaner state.